codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/endpoint/EndpointParamsGenerator.kt

```diff
@@ -0,0 +1,278 @@
+                        }
+                    },
+                )
+            }
+        }
+    }
+
+    private fun value(value: Value): String {
+        return when (value) {
+            is Value.String -> value.value().dq() + ".to_string()"
+            is Value.Bool -> value.expectBool().toString()
+            else -> TODO("unexpected type: $value")
+        }
+    }
+
+    private fun generateEndpointParamsBuilder(rustWriter: RustWriter) {
+        rustWriter.docs("Builder for [`Params`]")
+        Attribute.Derives(setOf(Debug, Default, PartialEq, Clone)).render(rustWriter)
+        rustWriter.rustBlock("pub struct Builder") {
+            parameters.toList().forEach { parameter ->
+                val name = parameter.memberName()
+                val type = parameter.symbol().makeOptional()
+                rust("$name: #T,", type)
+            }
+        }
+
+        rustWriter.rustBlock("impl Builder") {
+            docs("Consume this builder, creating [`Params`].")
+            rustBlockTemplate(
+                "pub fn build(self) -> Result<#{Params}, #{ParamsError}>",
+                "Params" to paramsStruct(),
+                "ParamsError" to paramsError(),
+            ) {
+                val params = writable {
+                    rustBlockTemplate("#{Params}", "Params" to paramsStruct()) {
+                        parameters.toList().forEach { parameter ->
+                            rust("${parameter.memberName()}: self.${parameter.memberName()}")
+                            parameter.default.orNull()?.also { default -> rust(".or(Some(${value(default)}))") }
+                            if (parameter.isRequired) {
+                                rustTemplate(
+                                    ".ok_or_else(||#{Error}::missing(${parameter.memberName().dq()}))?",
+                                    "Error" to paramsError(),
+                                )
+                            }
+                            rust(",")
+                        }
+                    }
+                }
+                rust("Ok(#W)", params)
+            }
+            parameters.toList().forEach { parameter ->
+                val name = parameter.memberName()
+                check(name == memberName(parameter.name.toString()))
+                check("set_$name" == setterName(parameter.name.toString()))
+                val type = parameter.symbol().mapRustType { t -> t.stripOuter<RustType.Option>() }
+                rustTemplate(
+                    """
+                    /// Sets the value for $name #{extraDocs:W}
+                    pub fn $name(mut self, value: #{type}) -> Self {
+                        self.$name = Some(value);
+                        self
+                    }
+
+                    /// Sets the value for $name #{extraDocs:W}
+                    pub fn set_$name(mut self, param: Option<impl Into<#{nonOptionalType}>>) -> Self {
```

Comment:
Isn't this backwards? I thought `fn $name` takes the `Into` impl while `fn set_$name` takes an `Option` without an `Into` in our builder pattern.

Reply:
ah yeah good catch